sklearn_numba_dpex/common/kernels.py

```diff
@@ -394,44 +320,52 @@ def partial_sum_reduction(
         # NB: This kernel only perform a partial reduction
         group_id = dpex.get_group_id(zero_idx)
         local_work_id = dpex.get_local_id(zero_idx)
-        first_work_id = local_work_id == zero_idx
 
-        size = summands.shape[minus_one_idx]
+        sum_axis_size = summands.shape[minus_one_idx]
+        n_work_group_per_row = result.shape[minus_one_idx]
 
-        local_values = dpex.local.array(local_values_size, dtype=dtype)
+        local_work_group_id_in_row = (group_id % n_work_group_per_row)
+        row_idx = group_id // n_work_group_per_row
 
-        first_value_idx = group_id * work_group_size * two_as_a_long
+        first_value_idx = local_work_group_id_in_row * reduction_block_size
         augend_idx = first_value_idx + local_work_id
         addend_idx = first_value_idx + work_group_size + local_work_id
 
+        local_values = dpex.local.array(local_values_size, dtype=dtype)
+
         # Each work item reads two value in global memory and sum it into the local
         # memory
-        if augend_idx >= size:
-            set_col_to_zero(local_values, local_work_id)
-        elif addend_idx >= size:
-            copy_col(summands, augend_idx, local_values, local_work_id)
+        if augend_idx >= sum_axis_size:
+            pass
+        elif addend_idx >= sum_axis_size:
+            local_values[local_work_id] = fused_unary_func_(
+                summands[row_idx, augend_idx]
+            )
         else:
-            add_cols(summands, augend_idx, addend_idx, local_values, local_work_id)
+            local_values[local_work_id] = fused_unary_func_(
+                summands[row_idx, augend_idx]
+            ) + fused_unary_func_(summands[row_idx, addend_idx])
```

Comment:
Does this read better?
```suggestion
            local_values[local_work_id] = (
                fused_unary_func_(summands[row_idx, augend_idx]) +
                fused_unary_func_(summands[row_idx, addend_idx])
            )
```